video: buffer size and NATIVE_SCREEN_WIDTH/_HEIGHT based on character columns/rows

logiclrd opened the pull request:
I noticed that `NATIVE_SCREEN_WIDTH` and `NATIVE_SCREEN_HEIGHT` were being redefined locally in a bunch of different places, always with the same values. This PR:

* Introduces macros `VGA_CHAR_WIDTH` and `VGA_CHAR_HEIGHT` that act as source of truth for the resolution of the text buffer.
* Consolidates macro definitions for `NATIVE_SCREEN_WIDTH` and `NATIVE_SCREEN_HEIGHT` into `vgamem.h` and redefines them to be based on `VGA_CHAR_WIDTH` and `VGA_CHAR_HEIGHT`.
* Updates all hardcoded 640x400 width/height references I could find to use `NATIVE_SCREEN_WIDTH` and `NATIVE_SCREEN_HEIGHT` instead.
* Updates a number of references to the text buffer's width and height to use `VGA_CHAR_WIDTH` and `VGA_CHAR_HEIGHT`. Updating everything comprehensively is a broader project that involves evaluating whether various UI elements are anchored to top/left or bottom/right.

logiclrd committed:
Consolidated macro definitions for NATIVE_SCREEN_WIDTH and NATIVE_SCREEN_HEIGHT into vgamem.h.
Updated all hardcoded 640x400 width/height references to use NATIVE_SCREEN_WIDTH and NATIVE_SCREEN_HEIGHT instead.

mrpapersonic commented on schism/vgamem.c:
```diff
@@ -142,7 +142,7 @@ static inline SCHISM_ALWAYS_INLINE int vgamem_unpack_halfw(int c)
 static uint32_t vgamem[4000] = {0};
```
vgamem size is relative to the screen size as well, so should be `(NATIVE_SCREEN_HEIGHT * NATIVE_SCREEN_WIDTH / 64)` where 64 is the size of one character, 8x8.

There should also be some static assertion in vgamem.h that makes sure the screen dimensions are both multiples of 8: `SCHISM_STATIC_ASSERT((NATIVE_SCREEN_WIDTH % 8 == 0) && (NATIVE_SCREEN_HEIGHT % 8 == 0), "screen dimensions are expected to be  multiples of 8")`

logiclrd replied:
Makes sense :-)

logiclrd replied:
I have added a commit that introduces source-of-truth macros `VGA_CHAR_WIDTH` and `VGA_CHAR_HEIGHT`, and makes `NATIVE_SCREEN_WIDTH` and `NATIVE_SCREEN_HEIGHT` be computed based on them.

There's immense scope for coordinate values to be computed based on `VGA_CHAR_WIDTH` and `VGA_CHAR_HEIGHT`, but in many cases it requires deciding whether coordinates are anchored to top/left or bottom-right. That's out of scope for this PR, in my opinion.

logiclrd committed:
Added macros VGAMEM_COLUMNS and VGAMEM_ROWS to vgamem.h and made NATIVE_SCREEN_WIDTH and NATIVE_SCREEN_HEIGHT be computed based on them.
Updated key_event_reset in main.c to use VGAMEM_COLUMNS and VGAMEM_ROWS instead of hardcoding 80 and 50.
Updated coordinate comparisons in page.c, page_help.c, page_info.c, page_loadmodule.c, page_timeinfo.c, page_vars.c, vgamem.c and video.c to use the macros instead of hardcoded 80x50 values.

mrpapersonic commented on include/vgamem.h:
```diff
@@ -24,6 +24,12 @@
 #ifndef SCHISM_VGAMEM_H_
 #define SCHISM_VGAMEM_H_
 
+#define VGA_CHAR_WIDTH          80
```
these names are kind of misleading (it suggests the width & height of a vga character, which is different than what these values actually are); I'd suggest `VGAMEM_TEXT_WIDTH` and `VGAMEM_TEXT_HEIGHT`, but I'm not sure if those are easier to understand without proper context. I'm open to other names as well!

logiclrd replied:
That's fair :-) What about `VGAMEM_COLUMNS` and `VGAMEM_ROWS`?

mrpapersonic replied:
That sounds good to me
